src/btree/bt_sync_obsolete.c

```diff
@@ -0,0 +1,291 @@
+      (void *)ref, obsolete ? "" : "not ", __wt_time_aggregate_to_string(&newest_ta, tp_string));
+
+    if (obsolete && ((ret = __wt_page_parent_modify_set(session, ref, true)) == 0)) {
+        __wt_verbose_debug2(session, WT_VERB_CHECKPOINT_CLEANUP,
+          "%p: marking obsolete disk page parent dirty", (void *)ref);
+        *ref_deleted = true;
+        WT_STAT_CONN_DATA_INCR(session, cc_pages_removed);
+        return (0);
+    }
+
+    return (ret);
+}
+
+/*
+ * __sync_obsolete_cleanup_one --
+ *     Check whether the ref is obsolete according to the newest stop time point and handle the
+ *     obsolete page by either removing it or marking it for urgent eviction. This code is a best
+ *     effort - it isn't necessary that all obsolete references are noticed and resolved
+ *     immediately. To that end some of the state checking takes the easy option if changes happen
+ *     between operations.
+ */
+static int
+__sync_obsolete_cleanup_one(WT_SESSION_IMPL *session, WT_REF *ref)
+{
+    WT_DECL_RET;
+    uint8_t new_state, previous_state;
+    bool busy, ref_deleted;
+
+    busy = ref_deleted = false;
+
+    /* Ignore root pages as they can never be deleted. */
+    if (__wt_ref_is_root(ref)) {
+        __wt_verbose_debug2(
+          session, WT_VERB_CHECKPOINT_CLEANUP, "%p: skipping root page", (void *)ref);
+        return (0);
+    }
+
+    /* Ignore internal pages, these are taken care of during reconciliation. */
+    if (F_ISSET(ref, WT_REF_FLAG_INTERNAL)) {
+        __wt_verbose_debug2(session, WT_VERB_CHECKPOINT_CLEANUP,
+          "%p: skipping internal page with parent: %p", (void *)ref, (void *)ref->home);
+        return (0);
+    }
+
+    /*
+     * Check in memory, deleted and on-disk pages for obsolescence. An initial state check is done
+     * without holding the ref locked - this is to avoid switching refs to locked if it's not
+     * worthwhile doing the check. It's possible that the ref changes state while we are doing these
+     * checks. That's OK - in the worst case we might not review the ref this time, but we will on
+     * subsequent reconciliations.
+     */
+    if (ref->state == WT_REF_DELETED || ref->state == WT_REF_DISK) {
+        WT_REF_LOCK(session, ref, &previous_state);
+        /*
+         * There are two possible outcomes from the subsequent checks:
+         * * The ref will be returned to it's previous state.
+         * * The ref will change from disk to deleted.
+         * Use a parameter to allow the functions to request a state change.
+         */
+        new_state = previous_state;
+        if (previous_state == WT_REF_DELETED)
+            ret = __sync_obsolete_deleted_cleanup(session, ref);
+        else if (previous_state == WT_REF_DISK) {
+            ret = __sync_obsolete_disk_cleanup(session, ref, &ref_deleted);
+            if (ref_deleted)
+                new_state = WT_REF_DELETED;
+            WT_REF_UNLOCK(ref, new_state);
+            WT_RET(ret);
```

Comment:
These two lines must be outside of the `else if` check so that the REF_UNLOCK applies to both states.